datadog.go

```diff
@@ -0,0 +1,102 @@
+package gzap
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/urfave/negroni"
+	"go.uber.org/zap/zapcore"
+)
+
+type Middleware struct{}
+
+
+func (m *Middleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	startTime := time.Now()
+	fields := []zapcore.Field{
+		String("http.method", r.Method),
+		String("http.url", r.URL.String()),
+	}
+
+	// We have myriad approaches to extract the 'true' client ip. We should try everything.
+	ip := r.Header.Get("Cf-Connecting-Ip")
+	if ip == "" {
+		ip = r.Header.Get("True-Client-Ip")
+		if ip == "" {
+			ip = r.Header.Get("X-Forwarded-For")
+			if ip != "" {
+				ips := strings.Split(ip, ",")
+				ip = strings.TrimSpace(ips[0])
+			} else {
+				ip = r.RemoteAddr
+			}
+		}
+	}
+
+	if ip != "" {
+		fields = append(fields, String("network.client.ip", ip))
+	}
+
+	userAgent := r.UserAgent()
+	if userAgent != "" {
+		fields = append(fields, String("http.useragent", userAgent))
+	}
+
+	referrer := r.Header.Get("Referer")
+	if referrer != "" {
+		fields = append(fields, String("http.referer", referrer))
+	}
+
+	request_id := r.Header.Get("X-Request-Id")
+  if request_id == "" {
+      request_id = r.Header.Get("X-Amzn-Trace-Id")
+	}
+	if request_id != "" {
+		fields = append(fields, String("http.request_id", request_id))
+	}
+
+	next(rw, r)
+
+	res := rw.(negroni.ResponseWriter)
+
+	statusCode := res.Status()
+	fields = append(fields, Int("http.status_code", statusCode))
+	fields = append(fields, Int64("network.bytes_read", r.ContentLength))
+
+	if res.Written() == true {
+		fields = append(fields, Int("network.bytes_written", res.Size()))
+	}
+
+
+	var responseLogger LevedLogger = logger.Info
+	if statusCode >= 400 && statusCode < 499 {
+		responseLogger = logger.Warn
+	} else if statusCode >= 500 && statusCode < 599 {
+		responseLogger = logger.Error
+	}
+
+	duration := time.Since(startTime).Nanoseconds()
+	fields = append(fields, Int64("duration", duration))
+
+	summary := fmt.Sprintf(
+		"%d %s %s (%s) %.2fms",
+		statusCode,
+		r.Method,
+		r.URL.RequestURI(),
+		ip,
+		float64(duration) / float64(1e6),
+	)
+
+	responseLogger(summary, fields...)
+}
+
+// DatadogLoggingHandler stores request-response information in attributes that are recognized
+// by datadog without extra transformations.
+func DatadogRequestLoggerHandler(handler http.Handler) http.Handler {
+	n := negroni.New()
+	n.Use(&Middleware{})
```

Comment:
Why bother constructing an empty named struct and assigning it here?

Reply:
Maybe I'm not familiar enough with go. How else would this be done? IIRC Negroni expects an interface of some sort for its middleware.

Reply:
Looking at the Godocs for that function it seems like it expects a handler for the middleware https://godoc.org/github.com/urfave/negroni#Negroni.Use, and it seems entirely optional. Tests pass when I remove it, so I don't think it's necessary. Have you tested this integrated against any applications yet or just with the unit tests?

Reply:
@ian-axelrod I would remove this and the Middlware interface unless you feel strongly about it

Reply:
That... doesn't make sense to me. If you don't add the middleware, how could it be used?

Reply:
You're not adding any middleware with that `Use` call though, with the following call you are. 

`UseHandler adds a http.Handler onto the middleware stack`

Reply:
I commented out the `Use` and it caused the tests to fail.

Reply:
```
root@99f2310eeabd:/go/src/gzap# go test -v ./...
=== RUN   TestDatadog
--- PASS: TestDatadog (0.00s)
=== RUN   TestGelfCore_Write
=== RUN   TestGelfCore_Write/GelfCore.Write_should_return_an_error_when_failing_to_send_message
=== RUN   TestGelfCore_Write/GelfCore.Write_should_return_nil_when_successfully_sending_a_log
--- PASS: TestGelfCore_Write (0.00s)
    --- PASS: TestGelfCore_Write/GelfCore.Write_should_return_an_error_when_failing_to_send_message (0.00s)
    --- PASS: TestGelfCore_Write/GelfCore.Write_should_return_nil_when_successfully_sending_a_log (0.00s)
=== RUN   TestInitLogger
=== RUN   TestInitLogger/InitLogger_should_return_a_noop_logger_when_running_a_test
=== RUN   TestInitLogger/InitLogger_should_return_a_dev_logger_when_no_GRAYLOG_HOST_is_set
=== RUN   TestInitLogger/InitLogger_should_return_a_json_logger_when_ENABLE_DATADOG_JSON_FORMATTER_is_set
--- PASS: TestInitLogger (0.00s)
    --- PASS: TestInitLogger/InitLogger_should_return_a_noop_logger_when_running_a_test (0.00s)
    --- PASS: TestInitLogger/InitLogger_should_return_a_dev_logger_when_no_GRAYLOG_HOST_is_set (0.00s)
    --- PASS: TestInitLogger/InitLogger_should_return_a_json_logger_when_ENABLE_DATADOG_JSON_FORMATTER_is_set (0.00s)
PASS
ok  	gzap	0.014s
root@99f2310eeabd:/go/src/gzap# go test -v ./...
=== RUN   TestDatadog
--- FAIL: TestDatadog (0.00s)
	datadog_test.go:15: Expected 1 - Got 0 
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 6 [running]:
testing.tRunner.func1(0xc42010c0f0)
	/usr/local/go/src/testing/testing.go:711 +0x2d2
panic(0x7d99e0, 0xa752d0)
	/usr/local/go/src/runtime/panic.go:491 +0x283
gzap.TestDatadog(0xc42010c0f0)
	/go/src/gzap/datadog_test.go:59 +0xc57
testing.tRunner(0xc42010c0f0, 0x864750)
	/usr/local/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x2de
exit status 2
FAIL	gzap	0.011s
```

Reply:
I think there is something else that is wonky here, maybe with the test/log recorder, cause UseHandler` calls `Use` it's in implementation https://sourcegraph.com/github.com/urfave/negroni/-/blob/negroni.go#L117 and I don't understand how adding an empty interface makes anything work. I couldn't figure it out quickly enough though and I don't want to block on this but it feels real off.